core/types.go

```diff
@@ -315,3 +316,36 @@ func (s ParSignedDataSet) Clone() (ParSignedDataSet, error) {
 
 	return resp, nil
 }
+
+// Slot is a beacon chain slot including chain metadata to infer epoch and next slot.
+type Slot struct {
+	Slot          int64
+	Time          time.Time
+	SlotDuration  time.Duration
+	SlotsPerEpoch int64
+}
+
+// Next returns the next slot.
+func (s Slot) Next() Slot {
+	return Slot{
+		Slot:          s.Slot + 1,
+		Time:          s.Time.Add(s.SlotDuration),
+		SlotsPerEpoch: s.SlotsPerEpoch,
+		SlotDuration:  s.SlotDuration,
+	}
+}
+
+// Epoch returns the epoch of the slot.
+func (s Slot) Epoch() int64 {
+	return s.Slot / s.SlotsPerEpoch
+}
+
+// LastInEpoch returns true if this is the last slot in the epoch.
+func (s Slot) LastInEpoch() bool {
+	return s.Slot%s.SlotsPerEpoch == s.SlotsPerEpoch-1
+}
+
+// FirstInEpoch returns true if this is the first slot in the epoch.
+func (s Slot) FirstInEpoch() bool {
+	return s.Slot%s.SlotsPerEpoch == 0
```

Comment:
do we assume that `s.Slot` is always a positive integer? coz this won't work for `-1`

Reply:
-1 would return false and that is correct no?

Reply:
yeah, it is correct. the testcase i had in my mind was epoch `-1`, ie, slots from `-32...-1`.

Reply:
not sure that is a valid test case